### PR TITLE
Add a set operation compiler.

### PR DESCRIPTION
**Pull request by riemanli**
This implements a set operation compiler which takes input as a set operation and decompose the cardinality computation of the set operation to a linear sum of cardinalities of union-sets.

------

This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)

### CODE REVIEW

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @chenweiw and @riemanli)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 34 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uFFVcBPLDIbgh2Wwg:-N6uFFVcBPLDIbgh2Wwh:b-ti04wf) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L34)):*
> ```Starlark
>     srcs = ["SetOperationCompiler.kt"],
>     deps = [
>         "//src/main/proto/wfa/measurement/internal/reporting:reporting_sets_service_kt_jvm_grpc_proto",
> ```

I don't see any code in here that depends on internal service methods?
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6u70h9EWBTg-kZ-gip:-N6u70h9EWBTg-kZ-giq:b3czlze) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L36)):*
> ```Starlark
>         "//src/main/proto/wfa/measurement/internal/reporting:reporting_sets_service_kt_jvm_grpc_proto",
>         "//src/main/proto/wfa/measurement/reporting/v1alpha:report_kt_jvm_proto",
>         "@wfa_common_jvm//src/main/kotlin/org/wfanet/measurement/common/grpc",
> ```

Only the top-level service impl should throw gRPC exceptions, so this library should not have this dependency.

Libraries can throw other exceptions that the gRPC service impl can handle to throw the appropriate gRPC exception instead.
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 156 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uCwq5Du0Pt7WcUN4f:-N6uCwq5Du0Pt7WcUN4g:b4cz26e) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L156)):*
> ```Kotlin
> 
>         launch {
>           convertSinglePrimitiveRegionToUnionSetCoefficientMap(
> ```

For the most part I'm trusting this code as I don't have as deep an understanding of the math.

The only comment I have is that we usually don't bother parallelizing operations unless the thread would be idle for some time (e.g. disk or network I/O). Whether it's useful otherwise sort of depends on overall CPU utilization. e.g. if the server isn't handling many requests, then parallelizing other operations may make the individual responses faster. If the CPU is already being utilized to handle a large number of requests, then parallelizing CPU-bound operations doesn't really help.

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @chenweiw and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 34 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uFFVcBPLDIbgh2Wwg:-N6usz3874O9O_EdDAEL:b2gsdbx) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L34)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I don't see any code in here that depends on internal service methods?
</blockquote></details>

You're right. That is a legacy code after I decided not to get internal service involved in this class. 
Done.
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6u70h9EWBTg-kZ-gip:-N6uujeV7uYwXPFAlvbG:b368ga) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L36)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Only the top-level service impl should throw gRPC exceptions, so this library should not have this dependency.

Libraries can throw other exceptions that the gRPC service impl can handle to throw the appropriate gRPC exception instead.
</blockquote></details>

Thanks for the suggestion. It makes great sense to me. 
Done.
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 156 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uCwq5Du0Pt7WcUN4f:-N6v6Fws7hpkc5Rfco08:b-aubd0w) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L156)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

For the most part I'm trusting this code as I don't have as deep an understanding of the math.

The only comment I have is that we usually don't bother parallelizing operations unless the thread would be idle for some time (e.g. disk or network I/O). Whether it's useful otherwise sort of depends on overall CPU utilization. e.g. if the server isn't handling many requests, then parallelizing other operations may make the individual responses faster. If the CPU is already being utilized to handle a large number of requests, then parallelizing CPU-bound operations doesn't really help.
</blockquote></details>

Based on our time complexity analysis, for a set operation with 20 reporting sets, in average it will take about an hour without the help from a memory cache. That's why I think utilizing parallel computing would be better. Please let me know if that's not the case here.

**Review by chenweiw — APPROVED** [review]
For most of the codes I can read them easily because they are a bit similar to C++. I focus on the math part of the codes, and as far as I can tell, the math implementation is correct. Thanks Rieman!

LGTM

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: all files reviewed, 3 unresolved discussions (waiting on @riemanli)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6u70h9EWBTg-kZ-gip:-N6xnGqVEObXb6bMpXZb:b-s9gvtz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L36)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Thanks for the suggestion. It makes great sense to me. 
Done.
</blockquote></details>

I still see the dependency here.
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 156 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uCwq5Du0Pt7WcUN4f:-N6xnOrH4p8bHqKbr2zu:b7r2h5s) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L156)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Based on our time complexity analysis, for a set operation with 20 reporting sets, in average it will take about an hour without the help from a memory cache. That's why I think utilizing parallel computing would be better. Please let me know if that's not the case here.
</blockquote></details>

I think in the end it will depend on the rate of requests to the reporting server. This is fine to leave for the moment, as I assume the product will have to gain a fair amount of success before a given reporting server instance is receiving a high number of requests per second. Therefore, responding to a single request more quickly is better.

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @chenweiw and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel` line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6u70h9EWBTg-kZ-gip:-N6xvKKpE_sdW2lY2Aoi:bht2xrn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/BUILD.bazel#L36)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I still see the dependency here.
</blockquote></details>

IntelliJ complained `kotlinx` is not recognized when I removed this dependency, so I left it here. Now, I understand it should be replaced by `@wfa_common_jvm//imports/kotlin/kotlinx/coroutines:core`. 

Thanks for catching it again.
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 156 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6uCwq5Du0Pt7WcUN4f:-N6xwgFx633wuSRwI4uK:brxp9sd) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/030c9f3b19dd8e98a22527a920e4599ef026b6c0/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L156)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I think in the end it will depend on the rate of requests to the reporting server. This is fine to leave for the moment, as I assume the product will have to gain a fair amount of success before a given reporting server instance is receiving a high number of requests per second. Therefore, responding to a single request more quickly is better.
</blockquote></details>

Sounds good!
___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 235 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6oL6-VL-uvixmED6gu-r2-235:-N6xwxdb-sWiA_73nYTh:b-ucfknh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4f9f49ac4e2d698f5d054b0fe6a59342b513db16/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L235)):*
<details><summary><i>Previously, chenweiw wrote…</i></summary><blockquote>

The changes in lines 234~235 are correct and really nice for avoiding race condition. 
</blockquote></details>

Thanks!

**Review by chenweiw — APPROVED** [review]
LGTM

**Review by kungfucraig — APPROVED** [review]
> Reviewed 3 of 4 files at r4, 1 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 2 of 5 files reviewed, 1 unresolved discussion (waiting on @chenweiw, @kungfucraig, and @SanjayVas)

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 2 of 5 files reviewed, 1 unresolved discussion (waiting on @chenweiw, @kungfucraig, and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt` line 235 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N6oL6-VL-uvixmED6gu-r2-235:-N7Wrswf2vZ6AoL6jdXh:b-26qjvt) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4f9f49ac4e2d698f5d054b0fe6a59342b513db16/src/main/kotlin/org/wfanet/measurement/reporting/service/api.v1alpha/SetOperationCompiler.kt#L235)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

Hi @chenweiw please resolve this discussion if it looks good to you. Thanks!

File: src/main/kotlin/org/wfanet/measurement/reporting/service/api/v1alpha/SetOperationCompiler.kt [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+private typealias NumberReportingSets = Int
+
+/** A mapping from a [UnionSet] to its coefficient in the Venn diagram region decomposition. */
+private typealias UnionSetCoefficientMap = Map<UnionSet, Int>
+
+/**
+ * A mapping for cardinality computation from a [PrimitiveRegion] to its decomposition in terms of
+ * union-sets represented by [UnionSetCoefficientMap]. Take a case of 3 reporting sets (rs0, rs1,
+ * rs2) as an example. A primitive region with its value equal to 3 (b'011') is the intersection of
+ * rs0 and rs1. The decomposition of the cardinality of the region =
```
Comment (chenweiw) [inline src/main/kotlin/org/wfanet/measurement/reporting/service/api/v1alpha/SetOperationCompiler.kt:54]: Following the lines 28~30, should it be the intersection of rs0, and rs1, and COMPLEMENT(rs2)?

**Review by chenweiw — CHANGES_REQUESTED** [review]
The added new comments looks good and much easier for the reader to understand, even though the algorithm itself is not easy to read. Only one comment right after the line 54 might need a tiny revision.

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: 2 of 5 files reviewed, 2 unresolved discussions (waiting on @chenweiw, @kungfucraig, and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v1alpha/SetOperationCompiler.kt` line 54 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624#-N7Nl05ovtq_oQsGUdTs-r7-54:-N7XfUEa6ZdIcuwFVX0W:bzapxhy) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/663971e922ce4880e78e32c93d758d245923ad25/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v1alpha/SetOperationCompiler.kt#L54)):*
<details><summary><i>Previously, chenweiw wrote…</i></summary><blockquote>

Following the lines 28~30, should it be the intersection of rs0, and rs1, and COMPLEMENT(rs2)?
</blockquote></details>

Thanks for catching that! Done.

**Review by chenweiw — APPROVED** [review]
LGTM

**Review by riemanli — COMMENTED** [review]
> Reviewed 1 of 4 files at r4, 1 of 4 files at r5, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/624)* status: all files reviewed, 2 unresolved discussions (waiting on @chenweiw and @SanjayVas)

**Pull request merged**